orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/job/PreconfiguredJobStage.java

```diff
@@ -0,0 +1,90 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.orca.clouddriver.pipeline.job;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.netflix.spinnaker.orca.clouddriver.config.PreconfiguredJobStageParameter;
+import com.netflix.spinnaker.orca.clouddriver.config.PreconfiguredJobStageProperties;
+import com.netflix.spinnaker.orca.clouddriver.exception.PreconfiguredJobNotFoundException;
+import com.netflix.spinnaker.orca.clouddriver.service.JobService;
+import com.netflix.spinnaker.orca.pipeline.TaskNode;
+import com.netflix.spinnaker.orca.pipeline.model.Stage;
+import groovy.util.Eval;
+import org.springframework.stereotype.Component;
+
+import java.util.Map;
+import java.util.Optional;
+
+@Component
+class PreconfiguredJobStage extends RunJobStage {
+
+  private JobService jobService;
+  private ObjectMapper objectMapper;
+
+  public PreconfiguredJobStage(Optional<JobService> optionalJobService) {
+    this.jobService = optionalJobService.orElse(null);
+    this.objectMapper = new ObjectMapper();
+  }
+
+  @Override
+  public void taskGraph(Stage stage, TaskNode.Builder builder) {
+    PreconfiguredJobStageProperties preconfiguredJob = jobService.getPreconfiguredStages().stream()
+      .filter( p -> stage.getType().equals(p.getType()))
+      .findFirst()
+      .orElseThrow(() -> new PreconfiguredJobNotFoundException(stage.getType()));
+
+    stage.setContext(overrideIfNotSetInContextAndOverrideDefault(stage.getContext(), preconfiguredJob));
+    super.taskGraph(stage, builder);
+  }
+
+  private Map<String, Object> overrideIfNotSetInContextAndOverrideDefault(Map<String, Object> context, PreconfiguredJobStageProperties preconfiguredJob) {
+    Map<String, Object> preconfiguredMap = objectMapper.convertValue(preconfiguredJob, Map.class);
+    preconfiguredJob.getOverridableFields().stream().forEach(field -> {
+        Object contextValue = context.containsKey(field) ? context.get(field) : null;
+        Object preconfiguredValue = preconfiguredMap.containsKey(field) ? preconfiguredMap.get(field) : null;
+      if (contextValue == null || preconfiguredValue != null) {
+        context.put(field, preconfiguredValue);
+      }
+    });
+
+    preconfiguredJob.getParameters().stream().forEach( parameter -> {
+      if (parameter.getDefaultValue() != null) {
+        Eval.xy(context, parameter.getDefaultValue(), String.format("x.%s = y.toString()", parameter.getMapping()));
```

Comment:
This (pre-existing) string `Eval` stuff is kind of scary and also a bit confusing.  I'm 99% sure this is just equivalent to the much safer and more readable:
```
context.put(parameter.getMapping(), parameter.getDefaultValue().toString())
```

Reply:
I'll experiment with it. If `Map.put` can do deeply nested keys that should be fine. `Eval.xy` allows us to replace things like `manifest.spec.template.spec.containers[0].env[0].value`.

Reply:
Oh, I had no idea that was supported by that `Eval`...yeah, that makes this more difficult.  If at all possible would be great to find a more robust solution.  (This would seem to allow things like assigning `metaClass` or other reserved things to the parameter map, which is not great.)

Reply:
Yeah, it's actually pretty handy because you can just define how a parameter maps to any of the fields. Could you provide an example for how you could assign `metaClass` using the things above?

Also, maybe we can just take the approach that preconfigured jobs are managed by the trusted operator? User supplied values are always `toString`ed so it seems safe to me.

Reply:
Artifact replacers in Clouddriver do something similar, however, they convert the map to a document and then use `JsonPath` to replace things.

Reply:
I dug a bit into this and don't think it's actually injectable from an end-user perspective as the uesr-supplied values are bound to expressions by `Eval.xy`...the inject is in the `parameter.getMapping()` which is operator-supplied as you say.

What I did run into while testing this out is that I got an NPE when I tried to assign more deeply nested maps unless the `context` already had the full structure of the map.  That is, if I start with an empty context and try:
```
Eval.xy(context, 'abc', "x.abc.def = y")
```
I get an NPE, but it's fine if I've already initialized `context = [abc: [:]]`.  Presumably something is already setting up the structure of the context before we get to this code?

I'd probably prefer something more explicit than eval of the string...but realize it's pre-existing and might be difficult to clean up.  To what extent are the new kubernetes jobs relying on this deeply-nested assignment behavior?

Reply:
Right, I guess it's assumed that you're mapping values directly onto an existing map which, in this case, is supplied by the preconfigured job.

Kubernetes jobs aren't relying on this _themselves_ but the preconfigured portion of them is (Titus included). This is used to allow the user to supply single values that are "overlayed" onto the configuration before it's executed by the providers run job interface.

Reply:
For posterity: The concern here is that you could craft a `mapping` which actually runs malicious code. While it's generally accepted that the operator trusted we're going to put some additional checks in place to ensure that mapping is actually a JSON path (as intended).